tracer/src/Datadog.Trace/Agent/MessagePack/SpanMessagePackFormatter.cs

```diff
@@ -189,9 +197,57 @@ private int Serialize(ref byte[] bytes, int offset, in SpanModel spanModel)
             offset += WriteTags(ref bytes, offset, in spanModel, tagProcessors);
             offset += WriteMetrics(ref bytes, offset, in spanModel, tagProcessors);
 
+            if (metaStructValues?.Count > 0)
+            {
+                WriteMetaStruct(ref bytes, ref offset, metaStructValues);
+            }
+
             return offset - originalOffset;
         }
 
+        private void WriteMetaStruct(ref byte[] bytes, ref int offset, IReadOnlyDictionary<string, object> metaStruct)
+        {
+            var data = new Dictionary<string, byte[]>();
+
+            // Since ConcurrentDictionary does not guarantee the order of the keys, we need to order them
+            // to make serialization deterministic.
+            // We need to know the size of the byte array before actually writing it. Depending on the size,
+            // the serializer will use a different amount of bytes in the header to encode the length of the byte array.
+            // That means that we need to know the size of the array, then write the header and then, the array itself.
+
+            foreach (var keyValuePair in metaStruct.OrderBy(keyValuePair => keyValuePair.Key, StringComparer.OrdinalIgnoreCase))
+            {
+                // 256 is the size that the serializer would reserve initially for empty arrays, so we create
+                // the buffer with that size to avoid this first resize. If a bigger size is required later, the serializer
+                // will resize it.
+                var buffer = new byte[256];
+
+                // PrimitiveObjectFormatter relies on MessagePackBinary
+                var bytesCopied = PrimitiveObjectFormatter.Instance.Serialize(ref buffer, 0, keyValuePair.Value, null);
```

Comment:
If we can serialize these directly into `bytes` it would be more efficient without creating and populating the `Dictionary<string, byte[]>` and iterating over the items a second time.

Reply:
I think `meta_struct` is meant for any raw bytes, and the backend code can decide how to handle them, so we shouldn't force the value to always be MessagePack here. Maybe the tags should be `<string, byte[]>` instead of `<string, object>`?

Reply:
That's an interesting point. I have defined meta_struct values now as arrays of bytes and moved the conversion from object to byte[] to a helper class. Thanks!